### PR TITLE
DNS Cookies 🍪

### PR DESCRIPTION
**Pull request by nils-wisiol**
Enable dnsdist packet cache and enable DNS Cookies (RFC 9018). Secrets need to be the same for servers with the same name, this will be a caveat for desec-automation. Configuration [is a bit weird](https://github.com/PowerDNS/pdns/issues/11810), let's see what pdns folks say to that.

Also, git and github seem to be fine with funky branch names (-:

### CODE REVIEW

File: dnsdist/conf/dnsdist.conf [hunk 1]
```diff
@@ -33,3 +33,11 @@ carbonServer(os.getenv('DESEC_NS_CARBONSERVER'), os.getenv('DESEC_NS_CARBONOURNA
 
 -- DoT
 addTLSLocal('0.0.0.0', '/etc/dnsdist/dot.cer', '/etc/dnsdist/dot.key')
+
+-- cache
+pc = newPacketCache(100000, {
+    maxTTL=86400, minTTL=0,
+    temporaryFailureTTL=60, staleTTL=60,
+    dontAge=false,
```
Comment (peterthomassen) [inline dnsdist/conf/dnsdist.conf]: [Should be true](https://dnsdist.org/reference/config.html#packetcache).

**Review by peterthomassen — CHANGES_REQUESTED** [review]
One nit.

Questions:

- Do we have data on the performance impact of the cache? (Our backend is lmdb, so from memory, too. We're mainly saving network calls, but duplicating data in memory.)
- How do you get to "about 80 MB of memory" for the cache?

File: dnsdist/conf/dnsdist.conf (hunk 1 above)
Comment (nils-wisiol) [reply dnsdist/conf/dnsdist.conf]: Good catch!

**Comment by nils-wisiol** [comment]
>     * Do we have data on the performance impact of the cache? (Our backend is lmdb, so from memory, too. We're mainly saving network calls, but duplicating data in memory.)

No. Best guess is that latency is improved by a tiny bit at the cost of some memory use.

>     * How do you get to "about 80 MB of memory" for the cache?

I took a conservative estimate from the dnsdist docs, but can't find it right now. With the current setting, assuming 512 byte per response you'd get 51MB.

The frontends nodes have currently 2.2GB of available memory on average (min: 2.1GB, max: 2.5GB).

**Comment by peterthomassen** [comment]
> > ```
> > * Do we have data on the performance impact of the cache? (Our backend is lmdb, so from memory, too. We're mainly saving network calls, but duplicating data in memory.)
> > ```
> 
> No. Best guess is that latency is improved by a tiny bit at the cost of some memory use.

Would it be worthwhile to do a performance test that compares qps under full load? (I think it would be good to know whether it goes down by 3% or by 70% -- how would be otherwise answer that?)

**Comment by nils-wisiol** [comment]
I do not think it is worthwhile because:

1. I am expecting the performance to go up and not down, as one hop is saved. (Namely the one between dnsdist and ns.)
2. Frontend load is currently very low.

**Comment by peterthomassen** [comment]
> * I am expecting the performance to go up and not down, as one hop is saved. (Namely the one between dnsdist and ns.)

Convinced!

**Review by peterthomassen — APPROVED** [review]

**Comment by nils-wisiol** [comment]
For reference from [RFC 7873](https://www.rfc-editor.org/rfc/rfc7873.html#section-6)

>   For simplicity, it is RECOMMENDED that the same Server Secret be used
>   by each DNS server in a set of anycast servers.  If there is limited
>   time skew in updating this secret in different anycast servers, this
>   can be handled by a server accepting requests containing a
>   Server Cookie based on either its old or new secret for the maximum
>   likely time period of such time skew (see also [Section 7.1](https://www.rfc-editor.org/rfc/rfc7873.html#section-7.1)).

**Pull request merged**